### PR TITLE
Inference: Add FlashInfer routed MXFP8 MoE with bounded decode capacity

### PR DESCRIPTION
**Pull request by mathemakitten**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Adds routed block-scale MXFP8 support to the FlashInfer inference backend, including a graph-safe bounded-capacity path for decode-only execution.

The PyTorch grouped GEMM path materializes per-expert inputs and pads each expert to a 128-row boundary. On Nano at EP4, only 16.7% of rows contain useful tokens, resulting in 5.98x more GEMM rows than required.

FlashInfer's routed MoE kernel consumes the routing result directly and avoids the
per-expert 128-row padding. However, using over the complete dynamic-inference dispatcher buffer also performs unnecessary work, with that buffer containing 65,536 rows even when a decode step uses only a small prefix.

This PR adds a configurable decode token capacity. For decode-only graphs, FlashInfer operates on a bounded prefix when the host-known EP-wide token bound fits. Mixed prefill/decode steps, prefill steps, and oversized decode steps fallback to the full dispatcher capacity.

The full graph buffer remains allocated for CUDA graph stability; the optimization avoids quantizing, routing, executing, and copying its inactive tail.

Results compared to torch grouped GEMM: 
- torch: 0.0623-0.0614 ms/token
- this: 0.0405-0.0397 ms/token
- 35.2% less time per token
- 1.54x end-to-end throughput
- 2.6x faster during the steady decode-only plateau
- aggregate time from 64.872 seconds -> 42.051 seconds

This optimization targets decode-heavy or disaggregated-decode workloads where the token bound is known.

**Correctness checks**: Bounded and full-capacity FlashInfer execution were bit-exact in eager and CUDA graph validation with the same number of routed MoE calls and 0 mismatched calls/elements, matching a bf16 top-1 token reference.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by yaox12 — APPROVED** [review]

**Comment by wdykas** [comment]
@mathemakitten  Do we know how this will affect log prob diff?

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 1]
```diff
@@ -820,6 +821,12 @@ def _dynamic_step_forward_logits(self, input_ids: Tensor, position_ids: Tensor):
             position_ids (Tensor): The position IDs.
         """
         context = self.inference_wrapped_model.inference_context
+        decode_token_upper_bound = (
+            context.max_requests
+            * (self.num_speculative_tokens + 1)
+            * self.model_config.expert_model_parallel_size
+        )
+        InferenceMode.set_decode_state(context.is_decode_only(), decode_token_upper_bound)
```
Comment (santhnm2) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py]: Does `decode_token_upper_bound` need to be set as a persistent variable? Can we just compute it where it's needed since it's a static function of the inference config?

Comment (mathemakitten) [reply megatron/core/inference/text_generation_controllers/text_generation_controller.py]: Good call, done, thanks!

**Review by santhnm2 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]
